### ISSUE TITLE
Http Cache incorrect response after edit file multiple times

### ISSUE THREAD

**Issue by bufferUnderrun** [chronoxor/NetCoreServer#175]
Hi,

I'm tracking down a strange behaviour with HttpSession and cache.

I setup a static file server which contains a test.html
```csharp
// Create a new HTTP server
var server = new HttpCacheServer(IPAddress.Any, port);
server.AddStaticContent(www, "/");
```

1. When updating the file content while server is running and requesting at new, the response is correctly updated. [OK]
2. When updating the file a second time, the response is blank (no response from server). [FAILED]

**Comment by bufferUnderrun** [comment]
that's crazy, now i'm not reproducing the issue all the time but in a few cases.
i'm still investigate...

**Comment by chronoxor** [comment]
Please investigate it. I'm on vacation till 1st of February.

**Comment by bufferUnderrun** [comment]
**First**, the `_watcher.Changed()` event is fired twice just after editing.

It seems to be a [well-known bug/feature](https://stackoverflow.com/a/1765094/3699737) due to the way the underlying editing tools is working. I've tested with vscode and sublimetext, they do atomic write, so event called twice. 

**Then**, i add some debug line in the console and see interesting things happened

FileCache.cs
```csharp
private bool InsertFileInternal(string path, string file, string key, TimeSpan timeout, InsertHandler handler)
        {
            try
            {
                key = key.Replace('\\', '/');
                file = file.Replace('\\', '/');
                
                // Load the cache file content
                var content = File.ReadAllBytes(file);
                if (!handler(this, key, content, timeout)) {
                    Console.WriteLine($"{DateTime.Now.ToString("hh:mm:ss.fffff")} !handler"); //(1)
                    return false;
                }

                using (new WriteLock(_lockEx))
                {
                    // Update entries by path map
                    _entriesByPath[path].Add(key);
                    Console.WriteLine($"{DateTime.Now.ToString("hh:mm:ss.fffff")} InsertFileInternal {file} {content}"); // (2)
                }

                return true;
            }
            catch (Exception) { return false; }
        }
```

In some case, the  WriteLine in (2) shows that `content` variable was empty and so updated cache entry. It explains why i see blank in my response (see my previous post).

**Finaly**,

i think the problem come from this line : 

```chsarp
var content = File.ReadAllBytes(file);
```
I wonder if `File.ReadAllBytes()` support multithread or is not able to read the file while it's still writing from outside

What do you think ?

**Comment by chronoxor** [comment]
File IO operations must be synchronized - insert into cache only when file written. Good way to both operations from the same thread.

**Comment by bufferUnderrun** [comment]
What do you mean ? it is not fixable ?

**Comment by bufferUnderrun** [comment]
very frustrating, so by going deep under i found what's wrong.

1. As already said, there is a bug in watcher so ReadAllBytes() sometime return empty and the cache is poison. BUT it should finaly works it a save many time the same files.

**So there is something else**

2. As you can see in the screen bellow, for the same static file, it creates 2 differents keys in entriesByPath. The first one, is created on server start and the second is created after watcher.onChange(). It adds a "/" on filename, so the Cache was update with a right content but not in the good key. That's why Cache.Find() never return the update bytes.

![image](https://user-images.githubusercontent.com/8216040/151722675-1e2045ab-dbac-432f-98ac-325a53bb1d30.png)

**By sanitized the key inside InsertFileInternal() the bug is fixed**. Please accept the attach PR.

Thanks

**Comment by chronoxor** [comment]
Thanks for investigations! Fixed in 6.2.0
